doc/ACL/SAI-Proposal-Exact-Match.md

```diff
@@ -0,0 +1,123 @@
+SAI Exact Match API Proposal
+============================
+
+Title       | SAI Exact Match Proposal
+------------|--------------------------------
+Authors     | Midhun Somasundaran, Meta Inc
+            | Shrikrishna Khare, Meta Inc
+            | Jai Kumar, Broadcom Inc
+Status      | Draft
+Type        | Standards track
+Created     | 02/02/2023
+SAI-Version | 1.12
+---------------------------------------------
+
+# Overview #
+ACLs are generally implemented in hardware using TCAMs which have limited depth. The limited capacity of TCAM tables can become bottleneck for use cases such as ACL based forwarding. Some ASICs have capability to do exact match hash table lookup in the forwarding path. This allows matching packets based on a set of parameters similar to ACL TCAM matching and to set packet parameters or actions similar to ACL actions. This document proposes enhancements to the ACL spec to support exact match.
+
+## Exact Match Use Cases ##
+The model proposed solves two use cases
+
+- Increase ACL scale using exact match
+- Provide flexibility for EM to work in tandem with TCAM. For example, a TCAM ACL stage can set a metadata as part of its action which can be used as a part of the match key in EM ACL stage or vice versa.
+
+## Exact Match Behavior ##
+- There is no concept of entry ordering or priority with exact match. All entries need to have the same priority. The implementation can reject entries if there is a mix of priority values.
+- Masking selected bits in an ACL entry is not supported with exact match since the table is generally implemented using hash lookup. Implementation can reject entry add requests if the mask is not all 1s.
+- If all bits in a field are not required to be part of the lookup key, the valid bits can be specified at table level. Every ACL entry in the table will use the valid bit specification to determine the lookup key.
+
+# ACL Specification Enhancements #
+The proposal includes following changes to the ACL spec.
+
+1. Define a new ACL stage named SAI_ACL_STAGE_POST_INGRESS which is placed after router table lookup in the pipeline. The new ACL stage allows setting of user metadata from SAI_ACL_STAGE_INGRESS and matching the user metadata in SAI_ACL_STAGE_POST_INGRESS.
```

Comment:
Do we have an updated pipeline diagram/doc?  Thanks.

Reply:
Updated the v5 version which has ACL details with the new stage

Reply:
Can we please have a snapshot of the diagram included in this doc? That'll convey the idea better.